# Store Sales — time-series forecasting, honest purge/embargo CV

Case: [Store Sales — Time Series Forecasting](https://www.kaggle.com/competitions/store-sales-time-series-forecasting)
(Corporación Favorita grocery sales: daily unit sales for 1,782 store×family
series, 2013–2017, scored on **RMSLE**).

This is the time-series showcase the first notebooks could not be. Forecasting
has one structure an i.i.d. CV destroys: **time**. A random fold trains on the
future and scores the past, so its error is fiction. The honest estimate uses
**time-series CV** with purge/embargo (`time=`), an expanding window that only
ever trains on the past, with test windows sized in whole **calendar weeks** —
exactly how the competition's forward forecast is scored.

Two more capabilities ride along here:

- **the metric**: there is no native RMSLE, but RMSE on a `log1p(sales)` target
  *is* RMSLE on the original scale — so a one-line target transform makes the
  library's `rmse` selection match the leaderboard exactly (`expm1` inverts it);
- **datetime + lag features**: day-of-week, payday, days-since-earthquake, and
  past-only autoregressive lags/rolls — the bulk of the signal, all built to
  look strictly backward so the honest CV stays honest.

> Data note: the official competition files (rules accepted). `DATE_FROM` keeps a
> recent window so the runs finish in reasonable time; lags are computed on the
> full history first. Live submission stays off.

In [1]:
import json
import logging
import os
import shutil
import subprocess
import time
import zipfile
from pathlib import Path

import numpy as np
import pandas as pd

from honestml import (
    AutoML,
    CVConfig,
    FeatureSelectionConfig,
    FEConfig,
    HPOConfig,
    render_report,
    save_run_report,
)

SEED = 42
DATE_FROM = "2017-01-01"  # model window; lags still use the full history before it
DATA = Path("data/store-sales")
RESULTS = Path("results/store-sales")
RESULTS.mkdir(parents=True, exist_ok=True)
KAGGLE = os.environ.get("KAGGLE_BIN") or shutil.which("kaggle") or str(Path.home() / ".local" / "bin" / "kaggle")

logging.basicConfig(format="%(levelname)s %(name)s: %(message)s")
logging.getLogger("honestml").setLevel(logging.INFO)

In [2]:
if not (DATA / "train.csv").exists():
    DATA.mkdir(parents=True, exist_ok=True)
    subprocess.run(
        [KAGGLE, "competitions", "download", "-c", "store-sales-time-series-forecasting", "-p", str(DATA)],
        check=True,
    )
    with zipfile.ZipFile(DATA / "store-sales-time-series-forecasting.zip") as z:
        z.extractall(DATA)

In [3]:
train = pd.read_csv(DATA / "train.csv", parse_dates=["date"])
stores = pd.read_csv(DATA / "stores.csv")
oil = pd.read_csv(DATA / "oil.csv", parse_dates=["date"])
holidays = pd.read_csv(DATA / "holidays_events.csv", parse_dates=["date"])

# national holidays actually observed (transferred ones did not happen on that date)
nat = set(holidays.loc[(holidays["locale"] == "National") & (~holidays["transferred"]), "date"])

df = train.merge(stores, on="store_nbr", how="left").merge(oil, on="date", how="left")
df = df.sort_values(["store_nbr", "family", "date"]).reset_index(drop=True)
df["dcoilwtico"] = df.groupby(["store_nbr", "family"])["dcoilwtico"].ffill().bfill()

# past-only autoregressive features on log1p(sales) (the forecast target), built per series
df["log_sales"] = np.log1p(df["sales"])
g = df.groupby(["store_nbr", "family"])["log_sales"]
for k in (7, 14, 28):
    df[f"lag_{k}"] = g.shift(k)
df["roll_mean_7"] = g.transform(lambda s: s.shift(1).rolling(7).mean())
df["roll_mean_28"] = g.transform(lambda s: s.shift(1).rolling(28).mean())

# calendar / datetime-delta features
df["dow"] = df["date"].dt.dayofweek
df["month"] = df["date"].dt.month
df["day"] = df["date"].dt.day
df["year"] = df["date"].dt.year
df["is_payday"] = ((df["day"] == 15) | df["date"].dt.is_month_end).astype(int)
df["days_since_quake"] = (df["date"] - pd.Timestamp("2016-04-16")).dt.days
df["is_national_holiday"] = df["date"].isin(nat).astype(int)

df = df[df["date"] >= pd.Timestamp(DATE_FROM)].sort_values("date").reset_index(drop=True)
for c in ("store_nbr", "family", "cluster"):
    df[c] = df[c].astype("string")  # treat the store/family/cluster ids as categoricals

y = df["log_sales"]  # rmse on log1p(sales) == RMSLE on sales
tval = df["date"].to_numpy()  # the CV time axis (datetime64, for calendar-period folds)
daily = int(df.groupby("date").size().median())  # rows per day (~ number of series)
X = df.drop(columns=["id", "date", "sales", "log_sales"])
print(f"rows: {len(X)}, features: {X.shape[1]}, daily series: {daily}, window: {DATE_FROM}..{df['date'].max().date()}")
X.head()

rows: 404514, features: 20, daily series: 1782, window: 2017-01-01..2017-08-15


,store_nbr,family,onpromotion,city,state,type,cluster,dcoilwtico,lag_7,lag_14,lag_28,roll_mean_7,roll_mean_28,dow,month,day,year,is_payday,days_since_quake,is_national_holiday
0,1,AUTOMOTIVE,0,Quito,Pichincha,D,13,53.75,1.945910,1.945910,2.302585,1.485281,1.295005,6,1,1,2017,0,260,0
1,37,CELEBRATION,0,Cuenca,Azuay,D,2,53.75,3.526361,3.091042,1.791759,2.827455,2.656193,6,1,1,2017,0,260,0
2,37,BREAD/BAKERY,0,Cuenca,Azuay,D,2,53.75,5.920656,6.365933,6.296577,6.194563,6.288084,6,1,1,2017,0,260,0
3,37,BOOKS,0,Cuenca,Azuay,D,2,53.75,1.386294,1.098612,0.693147,0.495105,0.785697,6,1,1,2017,0,260,0
4,37,BEVERAGES,0,Cuenca,Azuay,D,2,53.75,8.014997,8.117611,8.058960,7.999845,7.962306,6,1,1,2017,0,260,0


## Run A — naive i.i.d. CV (a random k-fold, ignoring time)

A plain random 5-fold k-fold + a random 20% holdout — the default most quick
baselines reach for. On a forecasting task it mixes time freely: folds train on
dates *later* than the rows they score. Whether that inflates the number depends
on how much the model leans on time-invariant signal (the autoregressive lags)
versus fold-specific patterns — the honest time-series run next door is the
check, and the verdict below reads the two against each other.

In [4]:
model_iid = AutoML(
    task="regression",
    metric="rmse",
    cv=CVConfig(outer_holdout=0.2),
    random_state=SEED,
)
t0 = time.perf_counter()
model_iid.fit(X, y)
print(f"naive i.i.d. fit took {(time.perf_counter() - t0) / 60:.1f} min")
rep_iid = model_iid.run_report_
sel_iid = next(e["score"] for e in rep_iid["leaderboard"] if e["model_id"] == rep_iid["winner"])
print(f"winner {rep_iid['winner']}  selection RMSLE {sel_iid:.4f}  holdout {rep_iid['holdout_score']:.4f}  optimism {sel_iid - rep_iid['holdout_score']:+.4f}")
pd.DataFrame(rep_iid["leaderboard"])

INFO honestml.adapters.reader: auto-typing column=dow dtype=Int32 role=categorical reason=low_cardinality_int


INFO honestml.adapters.reader: auto-typing column=month dtype=Int32 role=categorical reason=low_cardinality_int


INFO honestml.adapters.reader: auto-typing column=year dtype=Int32 role=ignore reason=constant


INFO honestml.adapters.reader: auto-typing column=is_payday dtype=Int64 role=categorical reason=low_cardinality_int


INFO honestml.adapters.reader: auto-typing column=is_national_holiday dtype=Int64 role=categorical reason=low_cardinality_int


INFO honestml: stage key=run stage=selection elapsed=471.5s


WARNING honestml.adapters.boosting: boosting 'lightgbm' trained without early stopping; leaderboard comparison may favor overfit settings


INFO honestml: stage key=run stage=refit elapsed=1.8s


INFO honestml: stage key=run stage=refit elapsed=2.0s


INFO honestml: stage key=run stage=finalize elapsed=2.0s


naive i.i.d. fit took 7.9 min
winner lightgbm  selection RMSLE 0.3771  holdout 0.3784  optimism -0.0013


,model_id,score,rank
0,lightgbm,0.377061,1
1,catboost,0.384285,2
2,xgboost,0.384929,3
3,linear,0.548179,4
4,baseline,2.548911,5


## Run B — honest time-series CV (calendar-week folds, purge/embargo)

`cv=CVConfig(scheme="timeseries_period", period="week", ...)` + `time=date`: the
window walks forward over whole **calendar weeks** instead of row counts. Each
fold tests `n_test=4` weeks and trains on all strictly earlier ones (expanding),
with a one-week **purge** gap straddling the train/test boundary and a one-week
**embargo** after each test block (de Prado). Sizing in calendar periods means
the split no longer leans on a hand-computed rows-per-day: `n_test`, `purge` and
`embargo` count **weeks** directly. The outer holdout is the **latest window**,
scored once. This is the number the leaderboard would actually give you.

In [5]:
model_ts = AutoML(
    task="regression",
    metric="rmse",
    cv=CVConfig(scheme="timeseries_period", period="week", n_splits=4, n_test=4, purge=1, embargo=1, outer_holdout=0.2),
    random_state=SEED,
)
t0 = time.perf_counter()
model_ts.fit(X, y, time=tval)
print(f"honest time-series fit took {(time.perf_counter() - t0) / 60:.1f} min")
rep_ts = model_ts.run_report_
sel_ts = next(e["score"] for e in rep_ts["leaderboard"] if e["model_id"] == rep_ts["winner"])
print(f"winner {rep_ts['winner']}  selection RMSLE {sel_ts:.4f}  holdout {rep_ts['holdout_score']:.4f}  optimism {sel_ts - rep_ts['holdout_score']:+.4f}")
print(f"cv: {rep_ts['cv']}")
print(f"naive i.i.d. holdout was {rep_iid['holdout_score']:.4f}; honest time-series holdout is {rep_ts['holdout_score']:.4f}")
pd.DataFrame(rep_ts["leaderboard"])

INFO honestml.adapters.reader: auto-typing column=dow dtype=Int32 role=categorical reason=low_cardinality_int


INFO honestml.adapters.reader: auto-typing column=month dtype=Int32 role=categorical reason=low_cardinality_int


INFO honestml.adapters.reader: auto-typing column=year dtype=Int32 role=ignore reason=constant


INFO honestml.adapters.reader: auto-typing column=is_payday dtype=Int64 role=categorical reason=low_cardinality_int


INFO honestml.adapters.reader: auto-typing column=is_national_holiday dtype=Int64 role=categorical reason=low_cardinality_int


INFO honestml: stage key=run stage=selection elapsed=49.3s


INFO honestml: stage key=run stage=refit elapsed=1.8s


INFO honestml: stage key=run stage=refit elapsed=2.0s


INFO honestml: stage key=run stage=finalize elapsed=2.0s


WARNING honestml.composition.facade: holdout (0.4082) scores 22% better than the winner's OOF (0.5203); the outer holdout may not be independent of DEV — suspect group-structured rows + target encoding leaking across a row-wise carve (finding #11)


honest time-series fit took 0.9 min
winner lightgbm  selection RMSLE 0.5203  holdout 0.4082  optimism +0.1121
cv: {'period': 'week', 'n_periods': 25, 'n_folds': 4, 'n_dropped_empty': 0}
naive i.i.d. holdout was 0.3784; honest time-series holdout is 0.4082


,model_id,score,rank
0,catboost,0.475398,1
1,lightgbm,0.520279,2
2,xgboost,0.526312,3
3,linear,0.578825,4
4,baseline,2.538540,5


## Run C — level 2: time-series CV + feature engineering + selection + HPO

The honest regime plus the level-2 machinery on the readable feature store:

- `FEConfig(frequency_encoding=True, intersections=True)` — frequency encoding
  for the high-cardinality `family`/`store_nbr`/city/state, and categorical
  **intersections** (store×family is the natural series key). Target encoding is
  binary-only, so it is gracefully skipped for this regression — the report says so;
- `FeatureSelectionConfig(strategy="importance", cutoff="auto")` — lag/rolling vs
  calendar vs static, with the honest no-selection gate;
- `preset="best"` + `HPOConfig` — significance-gated ensembling and Optuna tuning,
  all still under the expanding-window calendar-week time-series CV.

In [6]:
model_l2 = AutoML(
    task="regression",
    metric="rmse",
    preset="best",
    feature_engineering=FEConfig(frequency_encoding=True, intersections=True),
    feature_selection=FeatureSelectionConfig(strategy="importance", cutoff="auto"),
    hpo=HPOConfig(n_trials=15, timeout_s=600),
    cv=CVConfig(scheme="timeseries_period", period="week", n_splits=4, n_test=4, purge=1, embargo=1, outer_holdout=0.2),
    random_state=SEED,
)
t0 = time.perf_counter()
model_l2.fit(X, y, time=tval)
print(f"level-2 fit took {(time.perf_counter() - t0) / 60:.1f} min")

INFO honestml.composition.presets: preset best applied: ['ensemble']


INFO honestml.adapters.reader: auto-typing column=dow dtype=Int32 role=categorical reason=low_cardinality_int


INFO honestml.adapters.reader: auto-typing column=month dtype=Int32 role=categorical reason=low_cardinality_int


INFO honestml.adapters.reader: auto-typing column=year dtype=Int32 role=ignore reason=constant


INFO honestml.adapters.reader: auto-typing column=is_payday dtype=Int64 role=categorical reason=low_cardinality_int


INFO honestml.adapters.reader: auto-typing column=is_national_holiday dtype=Int64 role=categorical reason=low_cardinality_int


C:\Users\Admin\Documents\AutoML\.venv\lib\site-packages\tqdm\auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm


INFO honestml: stage key=run stage=hpo elapsed=746.8s


WARNING honestml.application.feature_selection: feature selection kept 5 of 74 features


WARNING honestml: feature selection (5 of 74 features) is not significantly better than no-selection and risks regressing; shipping all features (finding #10)


WARNING honestml: native categorical gate demoted 19 high-cardinality column(s) to ordinal codes: ['city__dow', 'city__family', 'city__month', 'cluster__dow', 'cluster__family', 'cluster__month', 'dow__family', 'dow__state', 'dow__store_nbr', 'family__is_national_holiday', 'family__is_payday', 'family__month', 'family__state', 'family__store_nbr', 'family__type', 'is_national_holiday__store_nbr', 'is_payday__store_nbr', 'month__state', 'month__store_nbr'] (native_cat_max_unique=64)


INFO honestml: stage key=run stage=selection elapsed=1813.0s


INFO honestml: stage key=run stage=ensemble elapsed=16.9s


WARNING honestml.adapters.boosting: boosting 'catboost' trained without early stopping; leaderboard comparison may favor overfit settings


INFO honestml: stage key=run stage=refit elapsed=47.7s


INFO honestml: stage key=run stage=refit elapsed=63.0s


INFO honestml: stage key=run stage=finalize elapsed=63.0s


WARNING honestml.composition.facade: holdout (0.3882) scores 16% better than the winner's OOF (0.4641); the outer holdout may not be independent of DEV — suspect group-structured rows + target encoding leaking across a row-wise carve (finding #11)


level-2 fit took 44.9 min


In [7]:
rep_l2 = model_l2.run_report_
sel_l2 = next(e["score"] for e in rep_l2["leaderboard"] if e["model_id"] == rep_l2["winner"])
print(f"winner          : {rep_l2['winner']}")
print(f"selection RMSLE : {sel_l2:.4f}   (time-series OOF; level 1 honest: {sel_ts:.4f})")
print(f"holdout RMSLE   : {rep_l2['holdout_score']:.4f}   (latest window; level 1 honest: {rep_ts['holdout_score']:.4f})")
print(f"optimism        : {sel_l2 - rep_l2['holdout_score']:+.4f}")
print()
print("fe      :", json.dumps(rep_l2["config"]["fe"], default=str)[:300])
print("fs      :", json.dumps(rep_l2["feature_selection"], default=str)[:400])
print("hpo     :", json.dumps(rep_l2["hpo"], default=str)[:400])
print("ensemble:", json.dumps(rep_l2["ensemble"], default=str)[:300])
save_run_report(rep_l2, RESULTS)
render_report(rep_l2, RESULTS, fmt="md")
pd.DataFrame(rep_l2["leaderboard"])

winner          : catboost
selection RMSLE : 0.4641   (time-series OOF; level 1 honest: 0.5203)
holdout RMSLE   : 0.3882   (latest window; level 1 honest: 0.4082)
optimism        : +0.0759

fe      : {"target_encoding": false, "te_smoothing": 10.0, "frequency_encoding": true, "intersections": true, "max_pairs": 50}
fs      : {"strategy": "importance", "n_selected": 74, "selected": ["onpromotion", "dcoilwtico", "lag_7", "lag_14", "lag_28", "roll_mean_7", "roll_mean_28", "day", "days_since_quake", "store_nbr_freq", "family_freq", "city_freq", "state_freq", "type_freq", "cluster_freq", "dow_freq", "month_freq", "is_payday_freq", "is_national_holiday_freq", "store_nbr", "family", "city", "state", "type", "cluster", "dow",
hpo     : {"backend": "optuna", "inner_cv": 3, "deterministic": false, "selection_oof_is_post_tuning": true, "tuned_on_full_feature_space": true, "cost_estimate_fits": 135, "tuned": {"catboost": {"chosen_params": {"depth": 10, "learning_rate": 0.06655140179588039, "iterat

,model_id,score,rank
0,catboost,0.464084,1
1,xgboost,0.539685,2
2,linear,0.600672,3
3,lightgbm,0.845653,4
4,baseline,2.538540,5


## What the three runs say

| regime | selection OOF RMSLE | holdout RMSLE |
| --- | --- | --- |
| naive i.i.d. | 0.3771 | 0.3784 |
| honest time-series (calendar-week) | 0.5203 | 0.4082 |
| time-series + level 2 | 0.4641 | 0.3882 |

**The leak lives in the selection number, and the time axis is what exposes it.**
The naive i.i.d. CV reports an out-of-fold RMSLE of **0.3771** — but that fold
mixes time, training on dates *later* than the rows it scores. The honest
calendar-week CV, walking forward and scoring only the future, reports **0.5203**:
the i.i.d. estimate understates the real forecasting error by ~0.14 RMSLE
(**~38%**). That is the number you would be fooled by if you selected a model on
shuffled folds. Note the forward OOF (0.5203) runs *above* the latest-window
holdout (0.4082): each expanding-window fold trains only on the weeks before it,
so the early, data-starved folds make the pooled OOF structurally conservative,
while the holdout is scored by a model refit on the full DEV history. On a short
~7-month series that gap is wide enough to trip a leakage heuristic
(`finding #11` in the log) — here a false alarm, the gap is the expanding-window
effect, not a leak — but the OOF stays the honest, conservative selection number.

**Level 2 lands right back on the base model — and the gates are why.** With
`frequency_encoding` + store×family `intersections`, `importance` feature
selection and Optuna HPO, the OOF nudges from 0.5203 to **0.4641** — but most of
that is just the equivalence band settling on **catboost** (already the strongest
single model at 0.4754 in level 1) with a small HPO lift, well inside catboost's
run-to-run noise. The heavy machinery earns nothing it can ship: **feature
selection** proposes a 5-of-74 subset that does not clear the no-selection gate,
so all 74 features ship (`finding #10`); and **Caruana ensembling** does *not*
ship — its candidate blend (linear 0.63, xgboost 0.26) is judged no better than
the single winner, so the gate keeps the plain catboost
(`ensemble.applied: false`). The shipped model is a plain catboost on the full
feature set; the report says, honestly, that the extra machinery did not earn its
keep here, before any leaderboard could.

Two capabilities landed cleanly along the way: **`log1p(sales)` makes `rmse` ==
RMSLE** (the leaderboard metric, exactly — `expm1` inverts the predictions), and
**target encoding was gracefully skipped** for the regression target
(`fe.target_encoding: false`) while frequency encoding and intersections still
applied — the library disabled the one binary-only transformer without failing
the run.